### Data Warehouse and data mining systems #2

In [1]:
import pandas as pd
import numpy as np 
import ast

DATA_DIR="datasets\\"

In [2]:
data1 = pd.read_csv("".join([DATA_DIR,"SpotifyTop200songs2017-2021\\data.csv"]), sep="#")
data1 = data1[data1['Track Name'].notna()] # deleting songs without names
data2 = pd.read_excel("".join([DATA_DIR,"SpotifyTop200songs2017-2021\Spotify 2010 - 2019 Top 100 Songs.xlsx"]))
data2["rank"] = np.tile(np.arange(1, 101), 10)
[i for i in range(1, 101)]*10
data2 = data2[data2['title'].notna()] # deleting songs without names
# print(data1.head(5))
# print('-'*120)
# print(data2.head(5))
print('-'*120)
data_merged = data1.merge(data2, left_on='Track Name', right_on='title', how='outer')
print(data_merged.shape)
print(data_merged.columns)
print(data_merged.head(10))
print(data_merged.isnull().values.any())

------------------------------------------------------------------------------------------------------------------------
(331286, 24)
Index(['Position', 'Track Name', 'Artist', 'Streams', 'Date', 'Genre', 'title',
       'artist', 'top genre', 'year released', 'added', 'bpm', 'nrgy', 'dnce',
       'dB', 'live', 'val', 'dur', 'acous', 'spch', 'pop', 'top year',
       'artist type', 'rank'],
      dtype='object')
   Position Track Name      Artist    Streams        Date  \
0       1.0    Starboy  The Weeknd  3135625.0  2017-01-01   
1       1.0    Starboy  The Weeknd  3342769.0  2017-01-02   
2       1.0    Starboy  The Weeknd  3563076.0  2017-01-03   
3       1.0    Starboy  The Weeknd  3619247.0  2017-01-04   
4       1.0    Starboy  The Weeknd  3579206.0  2017-01-05   
5       3.0    Starboy  The Weeknd  3590301.0  2017-01-06   
6       3.0    Starboy  The Weeknd  3460752.0  2017-01-07   
7       3.0    Starboy  The Weeknd  3127938.0  2017-01-08   
8       3.0    Starboy  The Weeknd


#









<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Position</th>
      <th>Track Name</th>
      <th>Artist</th>
      <th>Streams</th>
      <th>Date</th>
      <th>Genre</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>Starboy</td>
      <td>The Weeknd</td>
      <td>3135625</td>
      <td>2017-01-01</td>
      <td>['canadian pop', 'canadian contemporary r&amp;b', 'pop']</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>Closer</td>
      <td>The Chainsmokers</td>
      <td>3015525</td>
      <td>2017-01-01</td>
      <td>['pop', 'pop dance', 'tropical house', 'edm', 'electropop', 'dance pop']</td>
    </tr>
    <tr>
      <th>2</th>
      <td>3</td>
      <td>Let Me Love You</td>
      <td>DJ Snake</td>
      <td>2545384</td>
      <td>2017-01-01</td>
      <td>['pop', 'electronic trap', 'dance pop', 'edm', 'pop dance', 'pop rap']</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4</td>
      <td>Rockabye (feat. Sean Paul &amp; Anne-Marie)</td>
      <td>Clean Bandit</td>
      <td>2356604</td>
      <td>2017-01-01</td>
      <td>['pop', 'uk dance', 'dance pop', 'uk funky', 'tropical house', 'pop dance', 'post-teen pop', 'edm']</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5</td>
      <td>One Dance</td>
      <td>Drake</td>
      <td>2259887</td>
      <td>2017-01-01</td>
      <td>['toronto rap', 'canadian pop', 'canadian hip hop', 'rap', 'pop rap', 'hip hop']</td>
    </tr>
  </tbody>
</table>

In [3]:
merged = data_merged.loc[:,["title", "Track Name"]].isnull().values.all(axis=1)
null_indexes = [i for i in range(len(merged)) if merged[i]]
data_merged.loc[null_indexes,:].head(5) # our data is properly merged

Empty DataFrame
Columns: [Position, Track Name, Artist, Streams, Date, Genre, title, artist, top genre, year released, added, bpm, nrgy, dnce, dB, live, val, dur, acous, spch, pop, top year, artist type, rank]
Index: []

[0 rows x 24 columns]

In [4]:
merged = data_merged.loc[:,["Track Name"]].isnull().values.all(axis=1)
null_indexes = [i for i in range(len(merged)) if merged[i]]
not_null_indexes = [i for i in range(len(merged)) if not merged[i]]
data_merged.loc[null_indexes,:].head(5)

Position Track Name Artist  Streams Date Genre  \
330718       NaN        NaN    NaN      NaN  NaN   NaN   
330719       NaN        NaN    NaN      NaN  NaN   NaN   
330720       NaN        NaN    NaN      NaN  NaN   NaN   
330721       NaN        NaN    NaN      NaN  NaN   NaN   
330722       NaN        NaN    NaN      NaN  NaN   NaN   

                                                title      artist  \
330718                  STARSTRUKK (feat. Katy Perry)       3OH!3   
330719                    My First Kiss (feat. Ke$ha)       3OH!3   
330720                                I Need A Dollar  Aloe Blacc   
330721  Airplanes (feat. Hayley Williams of Paramore)       B.o.B   
330722              Nothin' on You (feat. Bruno Mars)       B.o.B   

          top genre  year released  ...   dB  live   val    dur  acous  spch  \
330718    dance pop         2009.0  ... -6.0  23.0  23.0  203.0    0.0   6.0   
330719    dance pop         2010.0  ... -4.0  36.0  83.0  192.0    1.0   8.0   
330720     pop soul         2010.0  ... -7.0   9.0  96.0  243.0   20.0   3.0   
330721  atl hip hop         2010.0  ... -4.0   4.0  38.0  180.0   11.0  12.0   
330722  atl hip hop         2010.0  ... -6.0   9.0  74.0  268.0   39.0   5.0   

         pop  top year  artist type  rank  
330718  70.0    2010.0          Duo   1.0  
330719  68.0    2010.0          Duo   2.0  
330720  72.0    2010.0         Solo   3.0  
330721  80.0    2010.0         Solo   4.0  
330722  79.0    2010.0         Solo   5.0  

[5 rows x 24 columns]

In [5]:
data_merged["Track Name"].fillna(data_merged.title, inplace=True)  
data_merged["Artist"].fillna(data_merged.artist, inplace=True)
data_merged["Date"].fillna(data_merged.added, inplace=True)
data_merged["Position"].fillna(data_merged["rank"], inplace=True)
data_merged.drop(columns=['title', 'artist', 'added', "rank"], inplace=True)
data_merged.loc[null_indexes,:].head(5)

Position                                     Track Name      Artist  \
330718       1.0                  STARSTRUKK (feat. Katy Perry)       3OH!3   
330719       2.0                    My First Kiss (feat. Ke$ha)       3OH!3   
330720       3.0                                I Need A Dollar  Aloe Blacc   
330721       4.0  Airplanes (feat. Hayley Williams of Paramore)       B.o.B   
330722       5.0              Nothin' on You (feat. Bruno Mars)       B.o.B   

        Streams        Date Genre    top genre  year released    bpm  nrgy  \
330718      NaN  2022‑02‑17   NaN    dance pop         2009.0  140.0  81.0   
330719      NaN  2022‑02‑17   NaN    dance pop         2010.0  138.0  89.0   
330720      NaN  2022‑02‑17   NaN     pop soul         2010.0   95.0  48.0   
330721      NaN  2022‑02‑17   NaN  atl hip hop         2010.0   93.0  87.0   
330722      NaN  2022‑02‑17   NaN  atl hip hop         2010.0  104.0  85.0   

        dnce   dB  live   val    dur  acous  spch   pop  top year artist type  
330718  61.0 -6.0  23.0  23.0  203.0    0.0   6.0  70.0    2010.0         Duo  
330719  68.0 -4.0  36.0  83.0  192.0    1.0   8.0  68.0    2010.0         Duo  
330720  84.0 -7.0   9.0  96.0  243.0   20.0   3.0  72.0    2010.0        Solo  
330721  66.0 -4.0   4.0  38.0  180.0   11.0  12.0  80.0    2010.0        Solo  
330722  69.0 -6.0   9.0  74.0  268.0   39.0   5.0  79.0    2010.0        Solo

In [6]:
data_merged.loc[null_indexes,:].shape

(568, 20)

In [7]:
genres = [[], [], []]

for genre_list, genre in zip(data_merged["Genre"], data_merged["top genre"]):
    if pd.isna(genre_list):
        genres[0].append(genre)
        genres[1].append(np.nan)
        genres[2].append(np.nan)
    else:
        genre_list = ast.literal_eval(genre_list)
        for i in range(min(len(genre_list), 3)):
            genres[i].append(genre_list[i])
        for i in range(len(genre_list), 3):
            genres[i].append(np.nan)

In [8]:
data_merged["genre_1"] = genres[0]
data_merged["genre_2"] = genres[1]
data_merged["genre_3"] = genres[2]

In [ ]:
# data_merged.rename(columns = {'old_col1':'new_col1', 'old_col2':'new_col2'}, inplace = True)

In [9]:
data_merged.drop(columns=["Genre", "top genre"], inplace=True)
data_merged.loc[null_indexes,:].head(5)

Position                                     Track Name      Artist  \
330718       1.0                  STARSTRUKK (feat. Katy Perry)       3OH!3   
330719       2.0                    My First Kiss (feat. Ke$ha)       3OH!3   
330720       3.0                                I Need A Dollar  Aloe Blacc   
330721       4.0  Airplanes (feat. Hayley Williams of Paramore)       B.o.B   
330722       5.0              Nothin' on You (feat. Bruno Mars)       B.o.B   

        Streams        Date  year released    bpm  nrgy  dnce   dB  ...   val  \
330718      NaN  2022‑02‑17         2009.0  140.0  81.0  61.0 -6.0  ...  23.0   
330719      NaN  2022‑02‑17         2010.0  138.0  89.0  68.0 -4.0  ...  83.0   
330720      NaN  2022‑02‑17         2010.0   95.0  48.0  84.0 -7.0  ...  96.0   
330721      NaN  2022‑02‑17         2010.0   93.0  87.0  66.0 -4.0  ...  38.0   
330722      NaN  2022‑02‑17         2010.0  104.0  85.0  69.0 -6.0  ...  74.0   

          dur  acous  spch   pop  top year  artist type      genre_1 genre_2  \
330718  203.0    0.0   6.0  70.0    2010.0          Duo    dance pop     NaN   
330719  192.0    1.0   8.0  68.0    2010.0          Duo    dance pop     NaN   
330720  243.0   20.0   3.0  72.0    2010.0         Solo     pop soul     NaN   
330721  180.0   11.0  12.0  80.0    2010.0         Solo  atl hip hop     NaN   
330722  268.0   39.0   5.0  79.0    2010.0         Solo  atl hip hop     NaN   

       genre_3  
330718     NaN  
330719     NaN  
330720     NaN  
330721     NaN  
330722     NaN  

[5 rows x 21 columns]